daemon/images/image_pull.go

```diff
@@ -124,3 +155,130 @@ func (i *ImageService) GetRepository(ctx context.Context, ref reference.Named, a
 	}
 	return repository, confirmedV2, lastError
 }
+
+type contentStoreForPull struct {
+	distribution.ContentStore
+	leases leases.Manager
+
+	mu       sync.Mutex
+	digested []digest.Digest
+}
+
+func (c *contentStoreForPull) addDigested(dgst digest.Digest) {
+	c.mu.Lock()
+	c.digested = append(c.digested, dgst)
+	c.mu.Unlock()
+}
+
+func (c *contentStoreForPull) getDigested() []digest.Digest {
+	c.mu.Lock()
+	digested := make([]digest.Digest, len(c.digested))
+	copy(digested, c.digested)
+	c.mu.Unlock()
+	return digested
+}
+
+func (c *contentStoreForPull) Writer(ctx context.Context, opts ...content.WriterOpt) (content.Writer, error) {
+	w, err := c.ContentStore.Writer(ctx, opts...)
+	if err != nil {
+		return nil, err
+	}
+	return &contentWriter{
+		cs:     c,
+		Writer: w,
+	}, nil
+}
+
+type contentWriter struct {
+	cs *contentStoreForPull
+	content.Writer
+}
+
+func (w *contentWriter) Commit(ctx context.Context, size int64, expected digest.Digest, opts ...content.Opt) error {
+	if err := w.Writer.Commit(ctx, size, expected, opts...); err != nil {
+		return err
+	}
+
+	w.cs.addDigested(expected)
+	return nil
+}
+
+func imageKey(id image.ID) string {
+	return "image-" + id.String()
+}
+
+type imageStoreWithLease struct {
+	image.Store
+	leases   leases.Manager
+	injested *contentStoreForPull
+	ctx      context.Context
+}
+
+func (s *imageStoreWithLease) Create(config []byte) (image.ID, error) {
+	id, err := s.Store.Create(config)
+	if err != nil {
+		return "", err
+	}
+
+	return id, s.updateLease(id)
+}
+
+func (s *imageStoreWithLease) Get(id image.ID) (*image.Image, error) {
```

Comment:
iiuc this `Get/Delete` is only used on pull and not in regular image operations. Would be good to leave comments about what conditions they actually get called.